benchmarks/benchmarks/sparse_linalg_svds.py

```diff
@@ -1,31 +1,62 @@
 import os
+import warnings
 import numpy as np
 from .common import Benchmark, safe_import
 
 with safe_import():
+    from scipy.linalg import svd
     from scipy.sparse.linalg import svds
 
+# ensure that we are benchmarking a consistent outcome;
+# (e.g. if the code wasn't able to find a solution accurately
+# enough the timing of the benchmark would become useless).
+msg = ("the benchmark code did not converge as expected, "
+       "the timing is therefore useless")
+
 
 class BenchSVDS(Benchmark):
-    # Benchmark SVD using the MatrixMarket test matrices recommended by the
+    # Benchmark SVDS using the MatrixMarket test matrices recommended by the
     # author of PROPACK at http://sun.stanford.edu/~rmunk/PROPACK/
+    # 'arpack' convergence uniformly for all singular values, while
+    # 'lobpcg' extreme singular values may converge much faster, so
+    # the assert is only evaluated on the top k/2 singular values.
+    # The assert uses the relative error since for some tested matrices
+    # the maximal singular values are very large due to poor matrix scaling.
+    # The `maxiter` and `tol` paremeters of `svds` are tuned for fair comaprison.
+    # The dense SVD solve is benchmaked as the base. 
     params = [
-        [25],
+        [20],
         ["abb313", "illc1033", "illc1850", "qh1484", "rbs480a", "tols4000",
          "well1033", "well1850", "west0479", "west2021"],
         # TODO: re-include propack
-        ['arpack', 'lobpcg']  # 'propack' failing (Aug. 2023)
+        ['arpack', 'lobpcg', 'svd']  # 'propack' failing (Aug. 2023)
     ]
     param_names = ['k', 'problem', 'solver']
 
-    def setup(self, k, problem, solver):
+    def __init__(self):
         dir_path = os.path.dirname(os.path.realpath(__file__))
         datafile = os.path.join(dir_path, "svds_benchmark_files",
                                 "svds_benchmark_files.npz")
-        matrices = np.load(datafile, allow_pickle=True)
-        self.A = matrices[problem][()]
+        self.matrices = np.load(datafile, allow_pickle=True)  
+
+    def setup(self, k, problem, solver):
+        if solver == 'svd' and problem == "tols4000":
+            # skip: fails breaking the benchmark
+            raise NotImplementedError()
+
+        self.A = self.matrices[problem][()]
+        _, s, _ = svd(self.A.toarray(), full_matrices=False)
+        self.top_singular_values = np.flip(s[:int(k/2)])
+        self.tol = k * np.max(self.A.shape) * np.finfo(float).eps
```

Comment:
I'm wondering about the relation between `tol=1e-6` in the benchmark and `self.tol` being set here based on `eps`.

AFAICT, this is saying roughly that even with a moderate `tol=1e-6`, we're expecting the _relative_ error per singular value to be on the order of eps? And then we set the cumulative tolerance by estimating the maximal singular value of A with its maximal entry (with a slack factor of 2 because we're only using the half the singular values), times the number of entries being compared (delta of relative accuracy of k'th singular value to 1)?

Sidenote: is it really the case that `svd` and `svds` have inverse ordering for the singular values? 🤦

Reply:
The `tol=1e-6` in  
```
                _, s, _ = svds(self.A, k=k, solver=solver, random_state=self.rng,
                               maxiter = 50, tol=1e-6)
```
that you probably refer to is difficult to meaningfully choose. 
The definition is to my recollection solver-dependent as `arpack` and `lobpcg` use very different stopping criteria.

The meaning of 
```
self.tol = k * np.max(self.A.shape) * np.finfo(float).eps
```
is fully controlled by the code of the benchmark and follows your kind of logic. I am unsure if this is the actual mathematically justified general definition, but it seems to work within the benchmark parameters.

Since the '`flip` is needed, yes, `svd` and `svds` have inverse orderings for the singular values.